# Memprediksi Emisi Gas CO2 beberapa Negara diseluruh Dunia 
### Projek machine learning
----------------------------------------

## Tahap 1: Data Cleaning and Preparation
-----------------------------------------

##### **Daftar Isi pada Notebook**
1. Background - Latar belakang pembuatan projek dan sumber data yang digunakan hingga tujuan penelitian
2. Notebook Setup - mengimport package dan data
3. Overview data
5. Data Cleaning
    - Menyesuaikan tipe data setiap variabel
    - Mengecek dan handle missing value
    - Me-*rename* variabel data
    - Menghapus kolom dan baris yang kosong
6. Transformasi DataFrame
    - *Melting* data pada setiap variabel
    - Mengintegrasikan data kedalam DataFrame yang sesuai
7. Export data yang telah dilakukan prepocessing kedalam format .csv
--------------------------------------------------------------------------

**1. Background**

**Latar Belakang Project**

Perubahan iklim di bumi kian terasa, salah satu alasan utam perubahan iklim dan suhu bumi adalah **emisi Co2**. Atmosfer bumi mengandung berbagai macam gas dari waktu ke waktu, beberapa gas dapat tertahan bertahun-tahun bahkan hingga ribuan tahun. Menurut survei yang dilakukan oleh *Inter Government Paneel* bahwasanya perubahan iklim suhu daratan dan lautan meningkat dimana rata-rata peningkatan suhu sebesar 0.85 C. Oleh karena itu emisi gas dalam tiga dekade terakhir meningkat yang membuat peningkatan suhu saat siang hari maupun malam hari [[1]](https://www.semanticscholar.org/paper/Prediction-Model%3A-CO2-Emission-Using-Machine-Kadam-Vijayumar/2f578a4b9ed92b51845df784dadfb6c8294f50ac). 

**Tujuan Project**

Pada Project Machine Learning ini akan menganalisis data berbagai negara dengan menggunakan model Machine Learning untuk memprediksi emisi gas CO2 dari beberapa feature. Dataset yang akan digunakan merupakan publik dataset perubahan iklim *World Bank Group* yang menyediakan data perubahan iklim di berbagai negara di dunia selama beberapa tahun dan beberapa feature yang ada didalamnya seperti:
- Negara : Sebagaian besar negara di seluruh duni dan kode negara
- Tahun  : Data yang disediakan mulai dari tahun 1990 hingga 2011
- Variasi Emisi Gas: CO2, CH4, N20, dan emisi gas yang lain
- Populasi : Jumlah populasi, pertumbuhan populasi, populasi di perkotaan, dan lainnya.
- Indikator Ekonomi Negara: GDP, GNI, Foreign Direct Investment, dan lainnya
- Lahan  : Hasil seral, lahan pertanian, kawasan terlindungi, dan lainnya
- Iklim  : Curah hujan, bencana nasional dan lainnya
- dan lainnya

Sebelum melakukan pemodelan machine learning, data yang akan digunakan harus melawati tahap prepocessing. Setiap tahapnya seperti pembersihan data, exploratory data analysis (EDA), dan analisis predeksi emisi gas CO2 dengan menggunakan Random Forest

--------------------------------------------------------------------------

**Sumber Data**

Dataset yang akan digunakan merupakan publik dataset perubahan iklim *World Bank Group* yang menyediakan data perubahan iklim di berbagai negara di dunia. Dapat diunduh pada [World Bank Group](https://datacatalog.worldbank.org/dataset/climate-change-data)

----------

**2. Notebook Setup**

Import semua package yang akan digunakan

In [1]:
import pandas as pd
import numpy as np
from functools import reduce

In [2]:
#mendefinisikan nama file dan sheet yang akan digunakan untuk proses analisis
orig_data = r"../Data/climate_change_download_0.xls"
sheet = "Data"

#meload data berekstensi excel menjadi pandas DataFrame
data_original = pd.read_excel(io=orig_data, sheet_name=sheet)

-------------------------------------------

**3. Overview Data**

Overview data ini untuk mengetahui beberapa data, tipe data, ukuran data, nama kolom/feature/variabel data sehingga mempermudah kita untuk mengeliminasi feature dan melakukan pemrosesan data

In [3]:
#Melihat data pada 2 baris pertama
data_original.head(2)

Country code Country name     Series code  \
0          ABW        Aruba  AG.LND.EL5M.ZS   
1          ADO      Andorra  AG.LND.EL5M.ZS   

                           Series name SCALE Decimals      1990 1991 1992  \
0  Land area below 5m (% of land area)     0        1  29.57481   ..   ..   
1  Land area below 5m (% of land area)     0        1         0   ..   ..   

  1993  ... 2002 2003 2004 2005 2006 2007 2008 2009 2010 2011  
0   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  
1   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  

[2 rows x 28 columns]

In [4]:
#Ukuran dataset
B, K = data_original.shape #tipe data berbentuk tuple 
print(f"Ukuran dataset yang kita miliki terdiri atas {B} Baris dan {K} Kolom")

Ukuran dataset yang kita miliki terdiri atas 13512 Baris dan 28 Kolom


In [5]:
#Nama-nama kolom
print("Nama-Nama Kolom pada Dataset")
data_original.columns

Nama-Nama Kolom pada Dataset


Index(['Country code', 'Country name',  'Series code',  'Series name',
              'SCALE',     'Decimals',           1990,           1991,
                 1992,           1993,           1994,           1995,
                 1996,           1997,           1998,           1999,
                 2000,           2001,           2002,           2003,
                 2004,           2005,           2006,           2007,
                 2008,           2009,           2010,           2011],
      dtype='object')

In [6]:
#Mengetahui banyaknya baris, kolom serta tipe data pada masing-masing feature
data_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13512 entries, 0 to 13511
Data columns (total 28 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Country code  13512 non-null  object
 1   Country name  13512 non-null  object
 2   Series code   13512 non-null  object
 3   Series name   13512 non-null  object
 4   SCALE         13512 non-null  object
 5   Decimals      13512 non-null  object
 6   1990          10017 non-null  object
 7   1991          10017 non-null  object
 8   1992          10017 non-null  object
 9   1993          10017 non-null  object
 10  1994          10017 non-null  object
 11  1995          10017 non-null  object
 12  1996          10017 non-null  object
 13  1997          10017 non-null  object
 14  1998          10017 non-null  object
 15  1999          10017 non-null  object
 16  2000          10017 non-null  object
 17  2001          10017 non-null  object
 18  2002          10017 non-null  object
 19  2003

In [7]:
#Melakukan statistika deskriptif pada semua feature
print("Statistika deskriptif pada semua feature: ")
data_original.describe()

Statistika deskriptif pada semua feature: 


Country code Country name     Series code  \
count         13512        13512           13512   
unique          233          233              58   
top             ABW        Aruba  AG.LND.EL5M.ZS   
freq             58           58             233   

                                Series name  SCALE  Decimals   1990   1991  \
count                                 13512  13512     13512  10017  10017   
unique                                   58      2         3   4355   3398   
top     Land area below 5m (% of land area)      0         1     ..     ..   
freq                                    233  10017      5823   5163   6520   

         1992   1993  ...   2002   2003   2004   2005   2006   2007   2008  \
count   10017  10017  ...  10017  10017  10017  10017  10017  10017  10017   
unique   3523   3583  ...   3877   3869   4007   4484   4008   4047   4080   
top        ..     ..  ...     ..     ..     ..     ..     ..     ..     ..   
freq     6364   6300  ...   5960   5974   5792   4933   5781   5769   5414   

         2009   2010   2011  
count   10017  10017  12382  
unique   3506   2164   1434  
top        ..     ..     ..  
freq     6256   7685  10244  

[4 rows x 28 columns]

Berdasarkan hasil statistika deskriptif diatas dapat diketahui ada 233 negara didalam dataset dan negara dengan frekuensi munculnya lebih sering adalah **Aruba**

Selanjutnya untuk memahami variabel *series code, series name, SCALE dan Decimals* maka diperlukan untuk mengetahui nilainya. Berikut ini kita akan menampilkan setiap nilai pada masing-masing variabel *series code, series name, SCALE dan Decimals*

In [8]:
#Nilai pada kolom Series Code
data_original['Series code'].unique()

array(['AG.LND.EL5M.ZS', 'AG.LND.IRIG.AG.ZS', 'AG.YLD.CREL.KG',
       'BX.KLT.DINV.WD.GD.ZS', 'EG.ELC.ACCS.ZS', 'EG.USE.COMM.GD.PP.KD',
       'EG.USE.PCAP.KG.OE', 'EN.ATM.CO2E.KT', 'EN.ATM.CO2E.PC',
       'EN.ATM.CO2E.PP.GD.KD', 'EN.ATM.GHGO.KT.CE', 'EN.ATM.METH.KT.CE',
       'EN.ATM.NOXE.KT.CE', 'EN.CLC.AERT', 'EN.CLC.DRSK.XQ',
       'EN.CLC.GHGR.MT.CE', 'EN.CLC.HCDM', 'EN.CLC.HJIP',
       'EN.CLC.HPPT.MM', 'EN.CLC.ICER', 'EN.CLC.IERU', 'EN.CLC.MDAT.ZS',
       'EN.CLC.MMDT.C', 'EN.CLC.NAMA', 'EN.CLC.NAPA', 'EN.CLC.NCOM',
       'EN.CLC.PCAT.C', 'EN.CLC.PCCC', 'EN.CLC.PCHW', 'EN.CLC.PCPT.MM',
       'EN.CLC.RNET', 'EN.POP.EL5M.ZS', 'EN.URB.MCTY.TL.ZS',
       'ER.H2O.FWTL.ZS', 'ER.LND.PTLD.ZS', 'IC.BUS.EASE.XQ',
       'IE.PPI.ENGY.CD', 'IE.PPI.TELE.CD', 'IE.PPI.TRAN.CD',
       'IE.PPI.WATR.CD', 'IQ.CPA.PUBS.XQ', 'IS.ROD.PAVE.ZS',
       'NY.GDP.MKTP.CD', 'NY.GNP.PCAP.CD', 'SE.ENR.PRSC.FM.ZS',
       'SE.PRM.CMPT.ZS', 'SH.DYN.MORT', 'SH.H2O.SAFE.ZS',
       'SH.MED.NUMW.P3', 'S

In [9]:
#Nilai pada kolom Series Name
data_original["Series name"].unique()

array(['Land area below 5m (% of land area)',
       'Agricultural land under irrigation (% of total ag. land)',
       'Cereal yield (kg per hectare)',
       'Foreign direct investment, net inflows (% of GDP)',
       'Access to electricity (% of total population)',
       'Energy use per units of GDP (kg oil eq./$1,000 of 2005 PPP $)',
       'Energy use per capita (kilograms of oil equivalent)',
       'CO2 emissions, total (KtCO2)',
       'CO2 emissions per capita (metric tons)',
       'CO2 emissions per units of GDP (kg/$1,000 of 2005 PPP $)',
       'Other GHG emissions, total (KtCO2e)',
       'Methane (CH4) emissions, total (KtCO2e)',
       'Nitrous oxide (N2O) emissions, total (KtCO2e)',
       'Annex-I emissions reduction target',
       'Disaster risk reduction progress score (1-5 scale; 5=best)',
       'GHG net emissions/removals by LUCF (MtCO2e)',
       'Hosted Clean Development Mechanism (CDM) projects',
       'Hosted Joint Implementation (JI) projects',
       'Av

In [10]:
#Nilai pada kolom SCALE
data_original["SCALE"].unique()

array([0, 'Text'], dtype=object)

In [11]:
#Nilai pada kolom Decimals
data_original["Decimals"].unique()

array([1, 0, 'Text'], dtype=object)

Namun, setelah kita mengetahui bahwa belum mengerti maksud nilai dari "Text" pada variabel SCALE dan Decimals maka dengan perintah berikut kita dapat lebih memahami sehingga mempermudah kita untuk membuat keputusan untuk menghilangkan datanya atau tetap menggunakan data tersebut

In [12]:
#Kita ingin mengetahui dataset pada kolom scale yang bernilai text
data_original[data_original["SCALE"]=="Text"]

Country code      Country name  Series code  \
3029          ABW             Aruba  EN.CLC.AERT   
3030          ADO           Andorra  EN.CLC.AERT   
3031          AFG       Afghanistan  EN.CLC.AERT   
3032          AGO            Angola  EN.CLC.AERT   
3033          ALB           Albania  EN.CLC.AERT   
...           ...               ...          ...   
7218          YEM       Yemen, Rep.  EN.CLC.RNET   
7219          ZAF      South Africa  EN.CLC.RNET   
7220          ZAR  Congo, Dem. Rep.  EN.CLC.RNET   
7221          ZMB            Zambia  EN.CLC.RNET   
7222          ZWE          Zimbabwe  EN.CLC.RNET   

                             Series name SCALE Decimals 1990 1991 1992 1993  \
3029  Annex-I emissions reduction target  Text     Text  NaN  NaN  NaN  NaN   
3030  Annex-I emissions reduction target  Text     Text  NaN  NaN  NaN  NaN   
3031  Annex-I emissions reduction target  Text     Text  NaN  NaN  NaN  NaN   
3032  Annex-I emissions reduction target  Text     Text  NaN  NaN  NaN  NaN   
3033  Annex-I emissions reduction target  Text     Text  NaN  NaN  NaN  NaN   
...                                  ...   ...      ...  ...  ...  ...  ...   
7218             Renewable energy target  Text     Text  NaN  NaN  NaN  NaN   
7219             Renewable energy target  Text     Text  NaN  NaN  NaN  NaN   
7220             Renewable energy target  Text     Text  NaN  NaN  NaN  NaN   
7221             Renewable energy target  Text     Text  NaN  NaN  NaN  NaN   
7222             Renewable energy target  Text     Text  NaN  NaN  NaN  NaN   

      ... 2002 2003 2004 2005 2006 2007 2008 2009 2010 2011  
3029  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3030  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3031  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3032  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3033  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  
7218  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   ..  
7219  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  Yes  
7220  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   ..  
7221  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   ..  
7222  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   ..  

[3495 rows x 28 columns]

In [13]:
#Kita ingin mengetahui dataset pada kolom scale yang bernilai text
data_original[data_original["Decimals"]=="Text"]

Country code      Country name  Series code  \
3029          ABW             Aruba  EN.CLC.AERT   
3030          ADO           Andorra  EN.CLC.AERT   
3031          AFG       Afghanistan  EN.CLC.AERT   
3032          AGO            Angola  EN.CLC.AERT   
3033          ALB           Albania  EN.CLC.AERT   
...           ...               ...          ...   
7218          YEM       Yemen, Rep.  EN.CLC.RNET   
7219          ZAF      South Africa  EN.CLC.RNET   
7220          ZAR  Congo, Dem. Rep.  EN.CLC.RNET   
7221          ZMB            Zambia  EN.CLC.RNET   
7222          ZWE          Zimbabwe  EN.CLC.RNET   

                             Series name SCALE Decimals 1990 1991 1992 1993  \
3029  Annex-I emissions reduction target  Text     Text  NaN  NaN  NaN  NaN   
3030  Annex-I emissions reduction target  Text     Text  NaN  NaN  NaN  NaN   
3031  Annex-I emissions reduction target  Text     Text  NaN  NaN  NaN  NaN   
3032  Annex-I emissions reduction target  Text     Text  NaN  NaN  NaN  NaN   
3033  Annex-I emissions reduction target  Text     Text  NaN  NaN  NaN  NaN   
...                                  ...   ...      ...  ...  ...  ...  ...   
7218             Renewable energy target  Text     Text  NaN  NaN  NaN  NaN   
7219             Renewable energy target  Text     Text  NaN  NaN  NaN  NaN   
7220             Renewable energy target  Text     Text  NaN  NaN  NaN  NaN   
7221             Renewable energy target  Text     Text  NaN  NaN  NaN  NaN   
7222             Renewable energy target  Text     Text  NaN  NaN  NaN  NaN   

      ... 2002 2003 2004 2005 2006 2007 2008 2009 2010 2011  
3029  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3030  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3031  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3032  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3033  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  
7218  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   ..  
7219  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  Yes  
7220  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   ..  
7221  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   ..  
7222  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   ..  

[3495 rows x 28 columns]

**Informasi yang kita dapatkan dari overview data**
- Ukuran dataset terdiri atas 13512 baris dan 28 kolom
- Pada misiing value sejumlah datanya bernilai Nan dan sebagai string ".."
- Kolom SCALE dan Decimals yang bernilai "Teks" tidak memiliki informasi apapun dan hampir seluruhnya terdiri dari nilai NaN

**4. Data Cleaning**

Pembersihan data dilakukan untuk mempermudahkita dalam melakukan proses pemodelan, seperti yang kita tahu istilah *"garbage in garbage out"* hasil yang kita inginkan bergantung dengan data yang seperti apa yang menjadi input. 

Pada proses cleaning ini kita akan  menghapus baris atau kolom yang kosong dan menangani beberapa kolom yang masih memiliki missing value.

Beberapa hal yang akan dilakukan pada proses ini antara lain:
1. Menghapus baris yang bernilai "Teks" pada variabel "SCALE" dan "Decimals"
2. Menghapus kolom yang tidak digunakan, kolom yang tidak digunakanan anatara lain:
    - Country code
    - Scale
    - Decimal
3. Mengubah string ".." dan "" menjadi nilai NaN untuk mempermudah mengenali nilai sebagai missing value
4. Mengubah nama fitur di variabel Series Name

In [14]:
#Kita buat variabel baru untuk memudahkan kita dalam memodifikasi DataFrame
cleaned_data = data_original

#menghapus baris yang bernilai "Text" pada kolom SCALE
cleaned_data = cleaned_data[cleaned_data["SCALE"]!="Text"]

In [15]:
#Mengetahui banyaknya baris baru
print("Banyaknya baris saat ini {}".format(cleaned_data.shape[0]))

Banyaknya baris saat ini 10017


In [16]:
#Menghapus feature yang tidak digunakan
#Beberapa feature yang tidak digunakan antara lain Country Code, Series Code, SCALE, dan Decimals
print("Banyaknya kolom saat ini:{}".format(cleaned_data.shape[1]))

cleaned_data = cleaned_data.drop(["Country code", "Series code", "SCALE",
                                  "Decimals"], axis=1)

print("Banyaknya kolom saat ini:{}".format(cleaned_data.shape[1]))

Banyaknya kolom saat ini:28
Banyaknya kolom saat ini:24


In [17]:
#Slicing DataFrame untuk semua baris dan dari index kolom ke 2 hingga akhir
cleaned_data.iloc[:,2:] = cleaned_data.iloc[:,2:].replace({"":np.nan, "..":np.nan})

In [18]:
data = cleaned_data.applymap(lambda x: pd.to_numeric(x, errors="ignore"))

In [19]:
#setelah melakukan beberapa transfromasi
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10017 entries, 0 to 13511
Data columns (total 24 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Country name  10017 non-null  object 
 1   Series name   10017 non-null  object 
 2   1990          4854 non-null   float64
 3   1991          3497 non-null   float64
 4   1992          3653 non-null   float64
 5   1993          3717 non-null   float64
 6   1994          3779 non-null   float64
 7   1995          4672 non-null   float64
 8   1996          3804 non-null   float64
 9   1997          3767 non-null   float64
 10  1998          3818 non-null   float64
 11  1999          4005 non-null   float64
 12  2000          5496 non-null   float64
 13  2001          4018 non-null   float64
 14  2002          4057 non-null   float64
 15  2003          4043 non-null   float64
 16  2004          4225 non-null   float64
 17  2005          5084 non-null   float64
 18  2006          4236 non-nul

In [20]:
data.head(5)

Country name                          Series name       1990  1991  1992  \
0        Aruba  Land area below 5m (% of land area)  29.574810   NaN   NaN   
1      Andorra  Land area below 5m (% of land area)   0.000000   NaN   NaN   
2  Afghanistan  Land area below 5m (% of land area)   0.000000   NaN   NaN   
3       Angola  Land area below 5m (% of land area)   0.208235   NaN   NaN   
4      Albania  Land area below 5m (% of land area)   4.967875   NaN   NaN   

   1993  1994  1995  1996  1997  ...  2002  2003  2004  2005  2006  2007  \
0   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
1   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
2   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
3   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
4   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   

   2008  2009  2010  2011  
0   NaN   NaN   NaN   NaN  
1   NaN   NaN   NaN   NaN  
2   NaN   NaN   NaN   NaN  
3   NaN   NaN   NaN   NaN  
4   NaN   NaN   NaN   NaN  

[5 rows x 24 columns]

| Variabel | Description | Unit |
| -------- | ----------- | ---- |
|land_area |Land area below 5m|% of land area|
|land_unIrrigation|Agricultural land under irrigation|% of total ag. land|
|cereal_yield|Cereal yield|kg per hectare|
|fdi_%GDP|Foreign direct investment, net inflows|% of GDP|
|acc_electricity_popul|Access to electricity|% of total population|
|energy_per_GDP|Energy use per units of GDP|kg oil|
|energy_per_capt|Energy use per capita|kilograms of oil equivalent|
|co2_total|CO2 emissions, total |KtCO2|
|co2_per_capt|CO2 emissions per capita |metric tons|
|co2_per_unit|CO2 emissions per units of GDP|kg per $1,000 of 2005 PPP $|
|other_ghg_emiss|Other GHG emissions, total|KtCO2e|
|ch4_emiss|Methane (CH4) emissions, total|KtCO2e|
|n20_emiss|Nitrous oxide (N2O) emissions, total |KtCO2e|
|disaster_risk|Disaster risk reduction progress score|1-5 scale; 5=best|
|ghg_nett|GHG net emissions/removals by LUCF|MtCO2e|
|DF_extreme|Droughts, floods, extreme temps|% pop. avg. 1990-2009|
|under_popul|Population below 5m|% of total|
|popul_agglomeration|Population in urban agglomerations >1million|%|
|freshwater|Annual freshwater withdrawals|% of internal resources|
|protected_areas|Nationally terrestrial protected areas|% of total land area|
|rate_business|Ease of doing business|ranking 1-183; 1=best|
|ie_private|Invest. in energy w/ private participation|$|
|itelecoms_private|Invest. in telecoms w/ private participation|$|
|itransport_private|Invest. in transport w/ private participation|$|
|iwater_private|Invest. in water/sanit. w/ private participation|$|
|rank_pub_sector|Public sector mgmt & institutions avg|1-6 scale; 6=best|
|paved_road|Paved roads|% of total roads|
|gdp|GDP|$|
|gni_capt|GNI per capita|Atlas $|
|ratio_g|Ratio of girls to boys in primary & secondary school|%|
|primary_rate|Primary completion rate, total|% of relevant age group|
|under_5_mortality|Under-five mortality rate|per 1,000|
|acc_improve_water|Access to improved water source|% of total pop|
|nurses_midwives|Nurses and midwives|per 1,000 people|
|physicians|Physicians|per 1,000 people|
|malaria_rate|Malaria incidence rate|per 100,000 people|
|acc_improve_sanit|Access to improved sanitation|% of total pop|
|underweight|Child malnutrition, underweight|% of under age 5|
|popul_living_below|Population living below $1.25 a day|% of total|
|popul_growth|Population growth|annual %|
|urb_popul_growth|Urban population growth|annual %|
|urb_popul|Urban population||

In [21]:
#Mendefiniskan kembali "series name" menjadi lebih relevan
rename_var = {'Land area below 5m (% of land area)': "land_area",
       'Agricultural land under irrigation (% of total ag. land)': "land_unIrrigation",
       'Cereal yield (kg per hectare)': "cereal_yield",
       'Foreign direct investment, net inflows (% of GDP)':"fdi_%GDP",
       'Access to electricity (% of total population)': "acc_electricity_popul",
       'Energy use per units of GDP (kg oil eq./$1,000 of 2005 PPP $)': "energy_per_GDP",
       'Energy use per capita (kilograms of oil equivalent)':"energy_per_capt",
       'CO2 emissions, total (KtCO2)':"co2_total",
       'CO2 emissions per capita (metric tons)':"co2_per_capt",
       'CO2 emissions per units of GDP (kg/$1,000 of 2005 PPP $)':"co2_per_unit",
       'Other GHG emissions, total (KtCO2e)':"other_ghg_emiss",
       'Methane (CH4) emissions, total (KtCO2e)':"ch4_emiss",
       'Nitrous oxide (N2O) emissions, total (KtCO2e)':"n20_emiss",
       'Disaster risk reduction progress score (1-5 scale; 5=best)':"disaster_risk",
       'GHG net emissions/removals by LUCF (MtCO2e)':"ghg_nett",
       'Droughts, floods, extreme temps (% pop. avg. 1990-2009)':"DF_extreme",
       'Population below 5m (% of total)':"under_popul",
       'Population in urban agglomerations >1million (%)':"popul_agglomeration",
       'Annual freshwater withdrawals (% of internal resources)':"freshwater",
       'Nationally terrestrial protected areas (% of total land area)':"protected_areas",
       'Ease of doing business (ranking 1-183; 1=best)':"rate_business",
       'Invest. in energy w/ private participation ($)':"ie_private",
       'Invest. in telecoms w/ private participation ($)':"itelecoms_private",
       'Invest. in transport w/ private participation ($)':"itransport_private",
       'Invest. in water/sanit. w/ private participation ($)':"iwater_private",
       'Public sector mgmt & institutions avg. (1-6 scale; 6=best)':"rank_pub_sector",
       'Paved roads (% of total roads)':"paved_road",
       'GDP ($)':"gdp",
       'GNI per capita (Atlas $)':"gni_capt",
       'Ratio of girls to boys in primary & secondary school (%)':"ratio_g",
       'Primary completion rate, total (% of relevant age group)':"primary_rate",
       'Under-five mortality rate (per 1,000)':"under_5_mortality",
       'Access to improved water source (% of total pop.)':"acc_improve_water",
       'Nurses and midwives (per 1,000 people)':"nurses_midwives",
       'Physicians (per 1,000 people)':"physicians",
       'Malaria incidence rate (per 100,000 people)':"malaria_rate",
       'Access to improved sanitation (% of total pop.)':"acc_improve_sanit",
       'Child malnutrition, underweight (% of under age 5)':"underweight",
       'Population living below $1.25 a day (% of total)':"popul_living_below",
       'Population growth (annual %)':"popul_growth", 
       'Urban population growth (annual %)':"urb_popul_growth",
       'Urban population':"urb_popul"}

In [22]:
data["Series name"] = data["Series name"].replace(to_replace=rename_var)

In [23]:
data.sample(10)

Country name        Series name           1990  \
1195                   Botswana     energy_per_GDP     131.305053   
702                      Angola           fdi_%GDP      -3.263096   
1555                Netherlands    energy_per_capt    4393.059724   
1503            Kyrgyz Republic    energy_per_capt    1692.557540   
13015  Turks and Caicos Islands         Population   11549.000000   
10140       Lower middle income           gni_capt     520.247265   
60                    Euro area          land_area       3.601821   
10908                    Serbia  under_5_mortality      28.600000   
12989                     Qatar         Population  473722.000000   
12861            Cayman Islands         Population   26048.000000   

                1991           1992           1993           1994  \
1195      124.292061     139.120239     136.944717     130.242972   
702         5.449515       4.982460       5.715529       4.195018   
1555     4602.581951    4503.489656    4511.785428    4509.968544   
1503     1509.992214    1095.324461     851.933744     615.472467   
13015   12216.000000   13005.000000   13848.000000   14644.000000   
10140     496.256279     493.636705     492.084521     507.572170   
60               NaN            NaN            NaN            NaN   
10908      26.600000      24.500000      22.600000      20.700000   
12989  482926.000000  488204.000000  491420.000000  495126.000000   
12861   27363.000000   28710.000000   30074.000000   31437.000000   

                1995           1996           1997  ...           2002  \
1195      127.665426     117.442621     111.853376  ...     101.419850   
702         9.374417       2.398994       5.363360  ...      14.626760   
1555     4576.176132    4732.306954    4580.545781  ...    4687.982280   
1503      519.437243     592.054327     548.837460  ...     499.809541   
13015   15333.000000   15859.000000   16275.000000  ...   23044.000000   
10140     553.837924     600.647861     617.715167  ...     607.114842   
60               NaN            NaN            NaN  ...            NaN   
10908      19.000000      17.400000      16.000000  ...      11.100000   
12989  501371.000000  512422.000000  528787.000000  ...  624173.000000   
12861   32792.000000   34096.000000   35361.000000  ...   45117.000000   

                2003           2004           2005           2006  \
1195       94.955478      87.194370      89.057108      86.212272   
702        25.112026       7.328525      -4.256403      -0.083508   
1555     4807.931279    4856.638946    4829.935573    4700.183242   
1503      533.501369     537.606017     516.836784     518.931261   
13015   25625.000000   28191.000000   30531.000000   32590.000000   
10140     680.357877     795.099044     913.628267    1038.464035   
60               NaN            NaN            NaN            NaN   
10908      10.500000       9.800000       9.300000       8.800000   
12989  653500.000000  715146.000000  820986.000000  978336.000000   
12861   47815.000000   50273.000000   52268.000000   53712.000000   

               2007          2008          2009          2010  2011  
1195   8.583384e+01  8.845891e+01  8.781111e+01           NaN   NaN  
702   -1.477781e+00  1.994536e+00  2.921200e+00  1.178046e+01   NaN  
1555   4.843796e+03  4.837167e+03  4.729160e+03  5.016127e+03   NaN  
1503   5.640632e+02  5.183353e+02  5.657467e+02           NaN   NaN  
13015  3.440400e+04  3.596000e+04  3.727100e+04  3.835400e+04   NaN  
10140  1.211221e+03  1.396561e+03  1.521726e+03  1.659717e+03   NaN  
60              NaN           NaN           NaN           NaN   NaN  
10908  8.300000e+00  7.800000e+00  7.400000e+00  7.100000e+00   NaN  
12989  1.178192e+06  1.396060e+06  1.597765e+06  1.758793e+06   NaN  
12861  5.467900e+04  5.529500e+04  5.576300e+04  5.623000e+04   NaN  

[10 rows x 24 columns]

In [24]:
#feature atau variabel yang akan digunakan
choosen_var = [
              "cereal_yield", "fdi_%GDP", "acc_electricity_popul", "energy_per_GDP", "energy_per_capt", "co2_total", 
              "co2_per_capt", "co2_per_unit", "other_ghg_emiss", "ch4_emiss", "n20_emiss", "DF_extreme",
              "popul_agglomeration", "protected_areas", "gdp","gni_capt", "under_5_mortality", "popul_growth", 
              "Population", "urb_popul_growth","urb_popul"
              ]

Kita akan mentransformasikan data, yang mana variabel series name yang telah kita pilih feature ddidalamnya menjadi kolom dan tahun menjadi satu kolom

In [25]:
# save the short feature names into a list of strings

# define an empty list, where sub-dataframes for each feature will be saved
frame_list = []

# iterate over all chosen features
for variable in choosen_var:
    
    # pick only rows corresponding to the current feature
    frame = data[data['Series name'] == variable]
    
    # melt all the values for all years into one column and rename the columns correspondingly
    frame = frame.melt(id_vars=['Country name', 'Series name']).rename(columns={'Country name': 'country', 'variable': 'year', 'value': variable}).drop(['Series name'], axis='columns')
    
    # add the melted dataframe for the current feature into the list
    frame_list.append(frame)


# merge all sub-frames into a single dataframe, making an outer binding on the key columns 'country','year'
from functools import reduce
all_vars = reduce(lambda left, right: pd.merge(left, right, on=['country','year'], how='outer'), frame_list)

In [26]:
all_vars

country  year  cereal_yield  fdi_%GDP  acc_electricity_popul  \
0                Aruba  1990           NaN       NaN                    NaN   
1              Andorra  1990           NaN       NaN                    NaN   
2          Afghanistan  1990        1200.6       NaN                    NaN   
3               Angola  1990         320.9 -3.263096                    NaN   
4              Albania  1990        2794.3       NaN                    NaN   
...                ...   ...           ...       ...                    ...   
5121       Yemen, Rep.  2011           NaN       NaN                    NaN   
5122      South Africa  2011           NaN       NaN                    NaN   
5123  Congo, Dem. Rep.  2011           NaN       NaN                    NaN   
5124            Zambia  2011           NaN       NaN                    NaN   
5125          Zimbabwe  2011           NaN       NaN                    NaN   

      energy_per_GDP  energy_per_capt  co2_total  co2_per_capt  co2_per_unit  \
0                NaN              NaN   1840.834     29.620641           NaN   
1                NaN              NaN        NaN           NaN           NaN   
2                NaN              NaN   2676.910      0.140715           NaN   
3         172.978279       569.242999   4429.736      0.428613    130.244398   
4         206.751128       809.215612   7488.014      2.276350    581.597641   
...              ...              ...        ...           ...           ...   
5121             NaN              NaN        NaN           NaN           NaN   
5122             NaN              NaN        NaN           NaN           NaN   
5123             NaN              NaN        NaN           NaN           NaN   
5124             NaN              NaN        NaN           NaN           NaN   
5125             NaN              NaN        NaN           NaN           NaN   

      ...  DF_extreme  popul_agglomeration  protected_areas           gdp  \
0     ...         NaN                  NaN         0.105547           NaN   
1     ...         NaN                  NaN         5.559145  1.028989e+09   
2     ...         NaN             6.739396         0.432993           NaN   
3     ...         NaN            15.171061        12.399821  1.026019e+10   
4     ...         NaN                  NaN         4.336615  2.101625e+09   
...   ...         ...                  ...              ...           ...   
5121  ...         NaN                  NaN              NaN           NaN   
5122  ...         NaN                  NaN              NaN           NaN   
5123  ...         NaN                  NaN              NaN           NaN   
5124  ...         NaN                  NaN              NaN           NaN   
5125  ...         NaN                  NaN              NaN           NaN   

      gni_capt  under_5_mortality  popul_growth  Population  urb_popul_growth  \
0          NaN                NaN      1.820254     62147.0          1.780501   
1      17440.0                8.8      2.945970     52773.0          2.840429   
2          NaN              208.7      2.082624  19023678.0          3.492189   
3        740.0              243.0      2.771320  10335052.0          6.505966   
4        680.0               41.1      1.027427   3289483.0          1.744276   
...        ...                ...           ...         ...               ...   
5121       NaN                NaN           NaN         NaN               NaN   
5122       NaN                NaN           NaN         NaN               NaN   
5123       NaN                NaN           NaN         NaN               NaN   
5124       NaN                NaN           NaN         NaN               NaN   
5125       NaN                NaN           NaN         NaN               NaN   

        urb_popul  
0       31259.941  
1       49976.031  
2     3481333.074  
3     3834304.292  
4     1197371.812  
...           ...  
5121          NaN  
5122          NaN  
5123          NaN  
5124   

In [27]:
#checking missing value
print("Check missing value pada setiap kolom:")
all_vars.isnull().sum()

Check missing value pada setiap kolom:


country                     0
year                        0
cereal_yield             1377
fdi_%GDP                 1111
acc_electricity_popul    5027
energy_per_GDP           2082
energy_per_capt          1956
co2_total                1143
co2_per_capt             1146
co2_per_unit             1557
other_ghg_emiss          4542
ch4_emiss                4526
n20_emiss                4526
DF_extreme               4958
popul_agglomeration      2582
protected_areas           726
gdp                       779
gni_capt                 1013
under_5_mortality         716
popul_growth              278
Population                252
urb_popul_growth          490
urb_popul                 467
dtype: int64

Untuk menghapus missing value kita akan menghapus sel kolom sebanyak mungkin serta mempertahankan jumlah kolom dan baris setinggi mungkin. Missing value dapat berasal dari data yang hilang pada variabel tahun, negara, atau fitur tertentu. Sehingga, diperlukan untuk memfilter data berdasarkan kolom atau baris dengan nilai NaN yang jumlah selnya hilang paling banyak. Dengan cara ini kita dapat membersihkan kumpulan data dan mempertahankan informasi maksimum yang mungkin berharga dari pada hanya melakukan penghapusan baris.

In [28]:
var_clean = all_vars

years_missing_value = dict.fromkeys(var_clean["year"].unique(), 0)
for index, row in var_clean.iterrows():
    years_missing_value[row["year"]] += row.isnull().sum()

years_missing_value_sort = dict(sorted(years_missing_value.items(), key=lambda item:item[1]))

print("missing value berdasarkan banyaknya missing value setiap tahunnya:")
for key, value in years_missing_value_sort.items():
    print(f"{key} : {value}")

missing value berdasarkan banyaknya missing value setiap tahunnya:
2005 : 1189
2000 : 1273
1995 : 1317
1990 : 1427
2007 : 1631
2006 : 1633
2004 : 1646
2008 : 1708
2003 : 1714
2002 : 1715
2001 : 1718
1999 : 1729
1998 : 1739
1997 : 1746
1996 : 1756
1994 : 1781
1993 : 1792
1992 : 1810
1991 : 1921
2009 : 2078
2010 : 3038
2011 : 4893


In [29]:
# filter only rows for years between 1991 and 2008 (having less missing values)
var_clean = var_clean[(var_clean['year'] >= 1991) & (var_clean['year'] <= 2008)]

print("Banyaknya missing value yang sebelum diurutkan berdasarkan tahun:")
print(var_clean.isnull().sum().sum())
print("Banyaknya baris setelah dilakukan pengurutan data berdasarkan tahun:")
print(var_clean.shape[0])

Banyaknya missing value yang sebelum diurutkan berdasarkan tahun:
29818
Banyaknya baris setelah dilakukan pengurutan data berdasarkan tahun:
4194


In [30]:
country_missing_value = dict.fromkeys(var_clean["country"].unique(), 0)
for index, row in var_clean.iterrows():
    country_missing_value[row["country"]] += row.isnull().sum()

country_missing_value_sort = dict(sorted(country_missing_value.items(), key=lambda item:item[1]))

print("missing value berdasarkan banyaknya missing value setiap tahunnya:")
for key, value in country_missing_value_sort.items():
    print(f"{key} : {value}")

missing value berdasarkan banyaknya missing value setiap tahunnya:
Angola : 81
Argentina : 81
Australia : 81
Austria : 81
Bangladesh : 81
Bulgaria : 81
Bolivia : 81
Brazil : 81
Canada : 81
Switzerland : 81
Chile : 81
China : 81
Cote d'Ivoire : 81
Cameroon : 81
Congo, Rep. : 81
Colombia : 81
Costa Rica : 81
Germany : 81
Denmark : 81
Dominican Republic : 81
Ecuador : 81
Egypt, Arab Rep. : 81
Euro area : 81
Spain : 81
Finland : 81
France : 81
United Kingdom : 81
Ghana : 81
Guatemala : 81
Honduras : 81
Hungary : 81
Indonesia : 81
India : 81
Ireland : 81
Israel : 81
Italy : 81
Jordan : 81
Japan : 81
Kenya : 81
Korea, Rep. : 81
Latin America & Caribbean : 81
Lower middle income : 81
Low & middle income : 81
Morocco : 81
Mexico : 81
Middle income : 81
Middle East & North Africa : 81
Mozambique : 81
Malaysia : 81
Nigeria : 81
Netherlands : 81
New Zealand : 81
Pakistan : 81
Panama : 81
Peru : 81
Philippines : 81
Portugal : 81
Paraguay : 81
Romania : 81
South Asia : 81
Saudi Arabia : 81
Sudan : 

In [31]:
#melakukan pemilihan data berdasarkan negara dimana hanya memilih baris data yang misiing valuenya kurang dari 90 
countries = []
for key, value in country_missing_value_sort.items():
    if value<90:
        countries.append(key)

var_clean = var_clean[var_clean["country"].isin(countries)]
print("Banyaknya missing value pada keseluruhan dataset yang sebelum diurutkan berdasarkan negara:")
print(var_clean.isnull().sum().sum())
print("Banyaknya baris setelah dilakukan pengurutan data berdasarkan negara:")
print(var_clean.shape[0])

Banyaknya missing value pada keseluruhan dataset yang sebelum diurutkan berdasarkan negara:
7854
Banyaknya baris setelah dilakukan pengurutan data berdasarkan negara:
1728


In [32]:
var_clean.isnull().sum()

country                     0
year                        0
cereal_yield               10
fdi_%GDP                   17
acc_electricity_popul    1728
energy_per_GDP              0
energy_per_capt             0
co2_total                   9
co2_per_capt                9
co2_per_unit                9
other_ghg_emiss          1446
ch4_emiss                1440
n20_emiss                1440
DF_extreme               1728
popul_agglomeration         0
protected_areas             0
gdp                         2
gni_capt                   16
under_5_mortality           0
popul_growth                0
Population                  0
urb_popul_growth            0
urb_popul                   0
dtype: int64

In [33]:
from itertools import compress

# create a boolean mapping of features with more than 20 missing values
bad_variabel = var_clean.isnull().sum()>20

# remove the columns corresponding to the mapping of the features with many missing values
cleans = var_clean.drop(compress(data = var_clean.columns, selectors = bad_variabel), axis=1)

print("Remaining missing values per column:")
print(cleans.isnull().sum())

Remaining missing values per column:
country                 0
year                    0
cereal_yield           10
fdi_%GDP               17
energy_per_GDP          0
energy_per_capt         0
co2_total               9
co2_per_capt            9
co2_per_unit            9
popul_agglomeration     0
protected_areas         0
gdp                     2
gni_capt               16
under_5_mortality       0
popul_growth            0
Population              0
urb_popul_growth        0
urb_popul               0
dtype: int64


In [34]:
# delete rows with any number of missing values
all_clean = cleans.dropna(axis='rows', how='any')

print("Remaining missing values per column:")
print(all_clean.isnull().sum())

print("Final shape of the cleaned dataset:")
print(all_clean.shape)

Remaining missing values per column:
country                0
year                   0
cereal_yield           0
fdi_%GDP               0
energy_per_GDP         0
energy_per_capt        0
co2_total              0
co2_per_capt           0
co2_per_unit           0
popul_agglomeration    0
protected_areas        0
gdp                    0
gni_capt               0
under_5_mortality      0
popul_growth           0
Population             0
urb_popul_growth       0
urb_popul              0
dtype: int64
Final shape of the cleaned dataset:
(1700, 18)


**6. Export Data**

In [35]:
#kita export dataset yang telah dicleaning kedalam format .csv
all_clean.to_csv("../Data/cleaned_data.csv", index=False)